sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystems.java

```diff
@@ -63,6 +78,71 @@
   /********************************** METHODS FOR CLIENT **********************************/
 
   /**
+   * This is the entry point to convert user-provided specs to {@link ResourceId ResourceIds}.
+   * Callers should use {@link #match} to resolve users specs ambiguities before
+   * calling other methods.
+   *
+   * <p>Implementation handles the following ambiguities of a user-provided spec:
+   * <ol>
+   * <li>{@code spec} could be a glob or a uri. {@link #match} should be able to tell and
+   * choose efficient implementations.
+   * <li>The user-provided {@code spec} might refer to files or directories. It is common that
+   * users that wish to indicate a directory will omit the trailing {@code /}, such as in a spec of
+   * {@code "/tmp/dir"}. The {@link FileSystem} should be able to recognize a directory with
+   * the trailing {@code /} omitted, but should always return a correct {@link ResourceId}
+   * (e.g., {@code "/tmp/dir/"} inside the returned {@link MatchResult}.
+   * </ol>
+   *
+   * <p>All {@link FileSystem} implementations should support glob in the final hierarchical path
+   * component of {@link ResourceId}. This allows SDK libraries to construct file system agnostic
+   * spec. {@link FileSystem FileSystems} can support additional patterns for user-provided specs.
+   *
+   * @return {@code List<MatchResult>} in the same order of the input specs.
+   *
+   * @throws IllegalArgumentException if specs are invalid -- empty or have different schemes.
+   * @throws IOException if all specs failed to match due to issues like:
+   * network connection, authorization.
+   * Exception for individual spec is deferred until callers retrieve
+   * metadata with {@link MatchResult#metadata()}.
+   */
+  public static List<MatchResult> match(List<String> specs) throws IOException {
+    checkArgument(!specs.isEmpty(), "Expect specs are not empty.");
+    Set<String> schemes = FluentIterable.from(specs)
```

Comment:
seems like it would be good to extract this logic to a common place -- "getOnlyScheme"? This would take a list of resources and assert that they all have the same scheme.

Note that we need this here for `String` but also for `ResourceId` in all the bulk APIs below.